lib/limiter/reservation.go

```diff
@@ -0,0 +1,133 @@
+	// will be returned.
+	//
+	// Implementations of ClientReserver are discouraged from blocking until
+	// a reservation becomes available.
+	TryReserve(ctx context.Context, c core.ClientID) (ClientReservation, error)
+
+	// ReleaseReservation releases a reservation that was previously acquired
+	// by TryReserve.
+	ReleaseReservation(ctx context.Context, r ClientReservation) error
+}
+
+// UnboundedClientReserver is a ClientReserver where all clients are free
+// to acquire arbitrarily many reservations without constraint.
+type UnboundedClientReserver struct{}
+
+func (u UnboundedClientReserver) TryReserve(ctx context.Context, c core.ClientID) (ClientReservation, error) {
+	return ClientReservation{c: c}, nil
+}
+
+func (u UnboundedClientReserver) ReleaseReservation(ctx context.Context, r ClientReservation) error {
+	return nil
+}
+
+// UniformlyBoundedClientReserver is a ClientReserver where all clients are
+// subject to a uniform maximum limit on the number of reservations they can
+// acquire at once.
+type UniformlyBoundedClientReserver struct {
+	MaxReservationsPerClient int64
+
+	// TODO consider also adding MaxConcurrentClients to bound amount of memory that
+	// resByClient map can consume. This could return a "reservations overloaded" error
+	// to signal to the caller that reservation system is currently overloaded.
+
+	// mu guards resByClient.
+	//
+	// TODO investigate fine-grain locking per ClientID to reduce lock contention.
+	// One naive idea is to use a two-level hierarchy of reservers where the top-level
+	// reserver hashes by ClientID to figure out which sub-reserver to delegate to.
+	// If the top-level reserver used a pre-specified pre-allocated fixed number of
+	// sub-reservers then that avoids the need to modify (and hence lock) the top
+	// level reserver at all, at the cost of creating a new parameter (fixed number
+	// of sub reservers) to tune to usage patterns.
+	//
+	// Alas, unlike various boring programming languages
+	//  https://doc.rust-lang.org/std/hash/index.html
+	//  https://docs.python.org/3/library/functions.html#hash
+	//  https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#hashCode--
+	// go doesn't provide a
+	// generic hash function for the application programmer to lean on. We could
+	// work around that by replacing ClientID with a concrete type or adding a
+	// Hash() uint64 method to the ClientID interface.
+	//
+	//
+	// See also: https://github.com/golang/go/issues/21035
+	mu          sync.Mutex
+	resByClient map[core.ClientID]int64
+}
+
+func NewUniformlyBoundedClientReserver(n int64) *UniformlyBoundedClientReserver {
+	return &UniformlyBoundedClientReserver{
+		MaxReservationsPerClient: n,
+		resByClient:              make(map[core.ClientID]int64),
+	}
+}
+
+func (b *UniformlyBoundedClientReserver) sanityCheck(n int64) {
+	// assert invariant 0 <= n <= MaxReservationsPerClient
+	if !((0 <= n) && (n <= b.MaxReservationsPerClient)) {
```

Comment:
```suggestion
	if n < 0 || n > b.MaxReservationsPerClient {
```

a bit easier to parse, IMO.

Reply:
De Morgan would approve